### ISSUE TITLE
Kube-proxy/ipvs;Frequently cannot simultaneous access clusterip and endpoint ip(real server) in pod

### ISSUE THREAD

**Issue by wangkai1994** [kubernetes/kubernetes#90042]
**What happened**:
I use calico cni and kube_proxy is on ipvs mode
![image](https://user-images.githubusercontent.com/9304834/78969395-6cfb9200-7b39-11ea-9159-8051fdcd7b42.png)

i found it's not stable when i simultaneous access apiserver in pod use clusterip and endpoint ip(real server)
 i find out:
** 172.31.0.1 is cluster ip
** 172.28.49.135 is pod ip
** 10.7.210.11 is endpoint ip (real server)
```
sh-4.2# cat /proc/net/nf_conntrack|grep 40008
ipv4     2 tcp      6 2 CLOSE src=172.28.49.135 dst=172.31.0.1 sport=40008 dport=443 src=172.31.0.1 dst=172.28.49.135 sport=443 dport=40008 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
ipv4     2 tcp      6 63 SYN_SENT src=172.28.49.135 dst=10.7.210.11 sport=40008 dport=16443 [UNREPLIED] src=10.7.210.11 dst=172.28.49.135 sport=16443 dport=40008 mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
```

**What you expected to happen**:

Stay stable when i simultaneous access apiserver in pod use clusterip and endpoint ip(real server)

**How to reproduce it (as minimally and precisely as possible)**:

enter an pod and execute two command:
1.while true; do curl -k https://{api-server-clusterip}:{api-server-service-port}/healthz; done
2.while true; do curl -k https://{api-server-realserver-ip}:{api-server-realserver-port}/healthz; done

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):1.18
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): CentOS 7
- Kernel (e.g. `uname -a`): 3.10.0-957.el7.x86_64
- Install tools:
- Network plugin and version (if this is a network-related bug): calico
- Others:

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by uablrek** [comment]
You havr the same combination of (new)K8s/(old)kernel as https://github.com/kubernetes/kubernetes/issues/89520. Please check if this is the same problem.

**Comment by wangkai1994** [comment]
@uablrek Oh it works ,thanks. But i don't find the root casue. Do you have any idea?

**Comment by uablrek** [comment]
How did you get it working? Upgrade kernel? Downgrade K8s?

The problem with ipvs+v1.18+linux-3.10 is that the lib used for controlling ipvs was upgraded but uses a feaure not present in linux-3.10. I think all things of bad things can happen and it's hard to know the symptoms in different systems.

However I think you should see error printouts in the `kube-proxy` log. I think this will be fixed in the next k8s v1.18.x. But I think it is a good idea in any case to switch to a newer kernel if you can.

**Comment by wangkai1994** [comment]
> Upgrade kernel

I upgrade kernel to 5.5 to solve this problem（but 4.4 is still not work,sad)

I think maybe this commit for kernel is the key?https://github.com/torvalds/linux/commit/3c5ab3f395d66a9e4e937fcfdf6ebc63894f028b

**Comment by uablrek** [comment]
The problem in https://github.com/kubernetes/kubernetes/issues/89520 only requires linux-4.x, so your problem the same, at least after upgrade to anything > linux-3.18.

**Comment by uablrek** [comment]
Is this a problem that has "popped up" after an upgrade? I mean, have is ever worked?

I believe this is some resource exaustion in your system. When you connect with the same src/dest in a tight-loop some resource limitation as likely to hit. But in your case the number of connect attempts ("ok"'s) seem not so many. A thing to check is if you have very many inactive connections in ipvs and/or very many sockets in TIME_WAIT in either of the endpoints.

**Comment by wangkai1994** [comment]
@uablrek 
it's worked when i upgrade kernel to 5.5.
And we found the "root cause"(not sure) probably is we set ip_local_reserved_ports in sysctl . We reserved about 10000 port in system, and kernel function of assign tcp port have some bug(always assign port which is already used) in 3.10.x but it be better in 4.x or 5.x  (not sure) .

**Comment by SerialVelocity** [comment]
This sounds a bit like something I've debugged before. Can you try setting `--local-port 40008` in `curl` for both loops?

The problem I know about is having two entries in the connection tracking table which are actually the same entry but there's a hidden DNAT.

e.g.
 - Conn1 connects to `172.31.0.1:443`
   - The connection on the client side is being tracked by the tuple `(172.28.49.135,40008,172.31.0.1,443)`
   - The connection on the server side is being tracked by the tuple `(172.28.49.135,40008,10.7.210.11,6443)`
 - Conn2 connects to `10.7.210.11:6443`
   - The connection on the client side is being tracked by the tuple `(172.28.49.135,40008,10.7.210.11,6443)`
   - The connection on the server side is being tracked by the tuple `(172.28.49.135,40008,10.7.210.11,6443)`

So, according to the above, the client has 2 distinct tuples being tracked while the server has only 1. This causes the server to combine both TCP streams into 1 causing a corrupted stream.

You should be able to check the connection tracking tables on both sides of the connection for confirmation when this happens.

**Comment by uablrek** [comment]
@SerialVelocity Thanks for the info. Good finding! With your input I am able to re-create the problem in a controlled way.

### Test service/pod

Alpine is used;
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: alpine-nc
spec:
  selector:
    matchLabels:
      app: alpine-nc
  replicas: 4
  template:
    metadata:
      labels:
        app: alpine-nc
    spec:
      containers:
      - name: alpine
        image: library/alpine:latest
        command: ["nc","-lk","-p","5001","-e","sh"]
---
apiVersion: v1
kind: Service
metadata:
  name: alpine-nc
spec:
  selector:
    app: alpine-nc
  ports:
  - name: nc
    port: 5001
```
You can login to a shell using `nc` through the service;
```
kubectl apply -f alpine-nc.yaml
nc alpine-nc.default.svc.cluster.local 5001
hostname
alpine-nc-84fff55cd9-zg7nw
```
You will not get any shell prompt, "hostname" is typed as input.

### Re-create the problem
```
$ kubectl get pods -l app=alpine-nc -o wide
NAME                         READY   STATUS    RESTARTS   AGE     IP         NODE     NOMINATED NODE   READINESS GATES
alpine-nc-84fff55cd9-h2tqk   1/1     Running   0          6m20s   11.0.0.3   vm-002   <none>           <none>
alpine-nc-84fff55cd9-nq484   1/1     Running   0          6m20s   11.0.2.2   vm-003   <none>           <none>
alpine-nc-84fff55cd9-rj6mv   1/1     Running   0          6m20s   11.0.1.2   vm-004   <none>           <none>
alpine-nc-84fff55cd9-zg7nw   1/1     Running   0          6m20s   11.0.3.2   vm-001   <none>           <none>
# Login to the pod in 3 different shells;
$ kubectl exec -it alpine-nc-84fff55cd9-h2tqk -- sh  # sh#1
$ kubectl exec -it alpine-nc-84fff55cd9-h2tqk -- sh  # sh#2
$ kubectl exec -it alpine-nc-84fff55cd9-h2tqk -- sh  # sh#3
# In sh#1
$ nc -p 48000 alpine-nc 5001
hostname
alpine-nc-84fff55cd9-zg7nw
# In sh#2
# Use the podip of the pod serving the request via clusterip
nc -p 48000 11.0.3.2 5001
(hangs...)
# In sh#3
$ netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      1 11.0.0.3:48000          11.0.3.2:5001           SYN_SENT    20/nc
tcp        0      0 11.0.0.3:48000          12.0.160.255:5001       ESTABLISHED 18/nc
tcp        0      0 :::5001                 :::*                    LISTEN      1/nc
```
### Using proxy-mode iptables

This problem does **not** happen with proxy-mode iptables.

I think the "random fully" options for NAT prevents the port crash. I don't think `ipvs` has that function. I fix may include a ipvs update (not likely to happen in the near future).

@wangkai1994 Using proxy-mode iptables is a work-around.

### Probability

There are many things that has to "go wrong" for this to happen. You must access a service direct and via clusterip simultaneously from the same pod using the same source port. For this to happen with any probability I think you must limit or tweak Linux ability to select ports.

I think this issue will affect *very* few users.

/remove-triage unresolved

**Comment by uablrek** [comment]
@aojea @thockin 

I think this problem can only be solved with some update in `ipvs` but should be *extremly* rare. It can be documented but is so obscure so I find it hard to describe in an understandable way.

I un-assign myself and let you decide how to proceed.

**Comment by uablrek** [comment]
This problem is not dependent on K8s or Linux kernel version. I used;
```
  17:44:11 Server Version: v1.18.2
  17:44:11 Linux 5.6.4 #2 SMP Thu Apr 16 18:52:41 CEST 2020
```

@wangkai1994 I think the reason you got better result with linux-5.x was better port randomization or pure luck.

**Comment by SerialVelocity** [comment]
@uablrek Can you try repro'ing this with iptables? If you connect using the cluster IP first, you then have the port to use when connecting directly to the pod. I'm wondering if the rules in iptables are considered before computing the conntrack entry.

**Comment by uablrek** [comment]
I did try with proxy-mode iptables and did not get the problem. 

> I think the "random fully" options for NAT prevents the port clash. I don't think ipvs has that function. A fix must probably include a ipvs update (not likely to happen in the near future), but can not be done in K8s as I see it.

**Comment by uablrek** [comment]
@wangkai1994 This;

> We reserved about 10000 port in system, and kernel function of assign tcp port have some bug(always assign port which is already used) in 3.10.x but it be better in 4.x or 5.x (not sure) .

is really interresting, specially since #90258 also apperas with a 3.10 kernel.

Do you have any more info on it?

**Comment by uablrek** [comment]
A test with proxy-mode=iptables;
```
In sh#1
nc -p 48000 alpine-nc 5001
hostname
alpine-nc-84fff55cd9-w7lp8
# in sh#2
$ nc -p 48000 11.0.2.2 5001
hostname
alpine-nc-84fff55cd9-w7lp8
# In sh#3 (client-pod)
$ netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 11.0.1.2:48000          12.0.240.90:5001        ESTABLISHED 10/nc
tcp        0      0 11.0.1.2:48000          11.0.2.2:5001           ESTABLISHED 16/nc
tcp        0      0 :::5001                 :::*                    LISTEN      1/nc
# In the server POD;
$ netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 :::5001                 :::*                    LISTEN      1/nc
tcp        0      0 ::ffff:11.0.2.2:5001    ::ffff:11.0.1.2:48000   ESTABLISHED 5/sh
tcp        0      0 ::ffff:11.0.2.2:5001    ::ffff:11.0.1.2:31801   ESTABLISHED 12/sh
```
The source port in the second connect is NAT'ed.

**Comment by uablrek** [comment]
I do not really understand what happens in iptables-mode above. Random-fully does not seem to have anything to do with it (my bad) since the source port 48000 is preserved in the connect to clusterip, but is NAT'ed in the direct connect. Is the kernel smart enough to see it has already used the source-port and NAT it even though it is a direct connect?

**Comment by SerialVelocity** [comment]
@uablrek Yes, you can "bypass" the "random fully" option by doing what I said above. You can look up the client port after you have initiated the cluster IP connection.

**Comment by uablrek** [comment]
Let's take this in steps so there is no misanderstanding;

1. Exec into a pod and connect to the clusterip explicitly using port 48000
```
$ nc -p 48000 alpine-nc 5001
hostname
alpine-nc-84fff55cd9-ps6zn
```
2. Check the connection in the *server* pod
```
$ netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 :::5001                 :::*                    LISTEN      1/nc
tcp        0      0 ::ffff:11.0.0.3:5001    ::ffff:11.0.3.2:48000   ESTABLISHED 5/sh
```
NOTE that the foreign port really is 48000, i.e it has *not* been NAT'ed
3. Make a direct connect to the server pod address from the same client pod again using port 48000
```
$ nc -p 48000 11.0.0.3 5001
hostname
alpine-nc-84fff55cd9-ps6zn
```
4. Check the connections in the client pod
```
$ netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 11.0.3.2:48000          11.0.0.3:5001           ESTABLISHED 20/nc
tcp        0      0 11.0.3.2:48000          12.0.10.171:5001        ESTABLISHED 19/nc
tcp        0      0 :::5001                 :::*                    LISTEN      1/nc
```
NOTE the local port for both the connect to clusterip and the direct connect (done *after* the connect to clusterip) uses the local port 4800
5. Check the connections on the server side
```
$ netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 :::5001                 :::*                    LISTEN      1/nc
tcp        0      0 ::ffff:11.0.0.3:5001    ::ffff:11.0.3.2:53226   ESTABLISHED 12/sh
tcp        0      0 ::ffff:11.0.0.3:5001    ::ffff:11.0.3.2:48000   ESTABLISHED 5/sh
```
Note that the foreign port of the direct connect is NAT'ed to 53226

**Comment by wangkai1994** [comment]
@uablrek Thanks for the troubleshooting! proxy-mode=iptables is good for us, but we want to found the root cause ,because we have the same problem when we use loadbalancer(based ipvs nat,like:keepalived) .

We think there is two problems:
1.IPVS NAT not support simultaneous access vip and realserver ip by same client port  ?
2.Realserver not support two connections use same client port   ?


@SerialVelocity 

> This sounds a bit like something I've debugged before. Can you try setting --local-port 40008 in curl for both loops?

curl may not support two client use the some local port.

![image](https://user-images.githubusercontent.com/9304834/79832896-f2563080-83dc-11ea-8531-14e4b0b434c3.png)

**Comment by andrewsykim** [comment]
Haven't read the whole thread here but it's very possible this is related to https://github.com/kubernetes/kubernetes/issues/89520 since you are using CentOS 7

**Comment by uablrek** [comment]
@andrewsykim That was my first thought also, but it was reported also in linux-4.x so it is different.

It looks like a connect via a service and a direct connect to the (same) endpoint done from the same client fails if the client port happens to be the same. This can be provoked (https://github.com/kubernetes/kubernetes/issues/90042#issuecomment-616660628) or it can happen by accident if you limit the number of port available for linux and connect like crazy both direct and via the service, which is what I think happens in this issue.

**Comment by andrewsykim** [comment]
Ah, seems related to https://github.com/kubernetes/kubernetes/issues/81775 then. I think this is largely a side effect of setting `conn_reuse_mode=0` and the way we do graceful termination which results in holding on to many TIME_WAIT connections. If https://github.com/kubernetes/enhancements/pull/1607 gets approved we could significantly reduce the # of TIME_WAIT entries because we can delete real servers immediately when the pod is deleted. In theory this should make this error happen significantly less but will have to do some test/validation to confirm.

**Comment by andrewsykim** [comment]
I think there are some cases where we want conn_reuse_mode=1, but the performance trade-offs are pretty significant and I suspect most users wouldn't find it worthwhile. However if there are users that can prove `conn_reuse_mode=1` _is_ better for their use-case we could start some discussions around adding a flag for it.

**Comment by SerialVelocity** [comment]
@andrewsykim That issue seems to be to do with high load. This can be reproduced with only two connections (where one connection is not using ipvs). Could you explain why this is to do with holding on to too many `TIME_WAIT` connections? (both connections are active in the repro mentioned on this issue)

**Comment by andrewsykim** [comment]
> @andrewsykim That issue seems to be to do with high load. 

There's more details in the end of the thread that are more relevant to this thread. 

> Could you explain why this is to do with holding on to too many TIME_WAIT connections? (both connections are active in the repro mentioned on this issue)

Generally we don't have to account for two connections from a pod using the same source port at the same time. However, you can have a stale connection (e.g. TIME_WAIT connection in conntrack) and a new connection try to use the same source port. This causes a conflict in the conntrack table. In the IPVS proxy because we don't delete stale real servers with TIME_WAIT entries in an attempt to gracefully terminate pods, the chances of port conflicts are much higher, especially on high load clusters.

**Comment by wangkai1994** [comment]
@andrewsykim Thank you. And one of my question is: Does IPVS NAT support simultaneous access vip and realserver ip by same source port ?

**Comment by andrewsykim** [comment]
Based on what @uablrek said it sounds like IPVS doesn't support it but I will try to validate it on my end this week and figure out the root cause for this behavior. Possibly a bug in the kernel or expected behavior.

**Comment by wangkai1994** [comment]
SNAT will assign the same port when connection network quintuple(s ip、d ip、policy、s port、d port) is same

**Comment by andrewsykim** [comment]
I thought the kernel will assign a new SNAT port if there's a collision. Is that not the case here?

**Comment by wangkai1994** [comment]
I read the kernel SNAT assign port code , i think kernel will not assign a new SNAT port if there's a collision when connection  network quintuple(s ip、d ip、policy、s port、d port) is same.That may caused the IPVS bug.But i hava no idea to fix it.

**Comment by SerialVelocity** [comment]
This might be a kernel issue rather than k8s issue

/remove-lifecycle rotten

**Comment by wangkai1994** [comment]
> rotten

Update:

**Comment by uablrek** [comment]
I think this is a duplicate of https://github.com/kubernetes/kubernetes/issues/81775. If that issue is fixed by PR https://github.com/kubernetes/kubernetes/pull/102122 I will close this issue also.
If possible, please try to re-create the problem in this issue with kernel >=5.9 and PR #102122 applied.

**Comment by wangkai1994** [comment]
@uablrek ok, thanks

**Comment by cyclinder** [comment]
Hi guys!

I reproduced this problem with kernel =5.9 and k8s=1.23. Client with same source port can't access to Virtual Server and Real Server at same time in IPVS mode. Please see shown the below:
```
[root@master ~]# uname -a
Linux master 5.9.8-1.el8.elrepo.x86_64 #1 SMP Wed Nov 11 09:27:50 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
[root@master ~]# kubectl version --short
Client Version: v1.23.6
Server Version: v1.23.7
[root@master ~]# kubectl get po -o wide
NAME     READY   STATUS    RESTARTS      AGE   IP               NODE     NOMINATED NODE   READINESS GATES
client   1/1     Running   1 (22h ago)   38h   192.168.219.77   master   <none>           <none>
ep       1/1     Running   0             18h   192.168.219.79   master   <none>           <none>
[root@master ~]# kubectl get svc
NAME         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
kubernetes   ClusterIP   10.96.0.1        <none>        443/TCP    16d
vip1         ClusterIP   10.108.122.180   <none>        8080/TCP   36h
```
the pod "ep" is RS of vip1.  
```bash
[root@master ~]# kubectl exec -it client bash
# client-bash1:
root@client:~# nc -p 5566 10.108.122.180 8080
```
Then will see the connection is **ESTABLISHED**:
```
[root@master ~]# conntrack -E -o ktimestamp -s 192.168.219.77
    [NEW] tcp      6 120 SYN_SENT src=192.168.219.77 dst=10.108.122.180 sport=5566 dport=8080 [UNREPLIED] src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=5566
 [UPDATE] tcp      6 60 SYN_RECV src=192.168.219.77 dst=10.108.122.180 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=5566
 [UPDATE] tcp      6 432000 ESTABLISHED src=192.168.219.77 dst=10.108.122.180 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=5566 [ASSURED]
```
and new a terminal for client-pod, and access to Real Server with same source port indirectly:
```bash
[root@master ~]# kubectl exec -it client bash
root@client:~# nc -p 5566 192.168.219.79 8080
Ncat: TIMEOUT.
```
Wait a while, Ncat reports **TIMEOUT**.  and look up conntrack entry on master：
```
[NEW] tcp 6 120 SYN_SENT src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 [UNREPLIED] src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=3663
[UPDATE] tcp 6 60 SYN_RECV src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=3663
[UPDATE] tcp 6 60 SYN_RECV src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=3663
[UPDATE] tcp 6 60 SYN_RECV src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=3663
[UPDATE] tcp 6 60 SYN_RECV src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=3663
[DESTROY] tcp 6 src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=3663
```
RS retry sends **SYN_RECV** four times but  no response, then the connection is destroyed. I notice that  the dport of reply conntrack is `3663`, Instead of the expected `5566`. I don't know why the port has changed, I checked the kernel source code and didn't find the logic to modify the port (maybe I missed it).

On client side(client pod),  The status of the connection stays at **SYN_SENT**：
```bash
root@client:~# netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      1 192.168.219.77:5566     192.168.219.79:8080     SYN_SENT    9615/nc
tcp        0      0 192.168.219.77:5566     10.108.122.180:8080     ESTABLISHED 9602/nc
tcp6       0      0 :::8080                 :::*                    LISTEN      1/./usr/bin/main
```

On server side(RS), The status of the connection stays at **SYN_RECV**：
```bash
root@ep:~# netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address Foreign Address State PID/Program name
tcp6 0 0 :::8080 :::* LISTEN 1/./usr/bin/main
tcp 0 0 192.168.219.79:8080 192.168.219.77:3663 SYN_RECV -
tcp 0 0 192.168.219.79:8080 192.168.219.77:5566 ESTABLISHED 1/./usr/bin/main
```  
I try to find some message about this by enable `net.netfilter.nf_conntrack_log_invalid`:
```
[root@master ~]# sysctl -w net.netfilter.nf_conntrack_log_invalid=6
net.netfilter.nf_conntrack_log_invalid = 6
```
I found out that kernel thinks this reply packet is not invalid, so drop it.
```
[94416.421511] nf_ct_proto_6: invalid packet ignored in state SYN_RECV  IN= OUT= SRC=192.168.219.77 DST=192.168.219.79 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=7877 DF PROTO=TCP SPT=5566 DPT=8080 SEQ=4068803588 ACK=0 WINDOW=64860 RES=0x00 SYN URGP=0 OPT (020405820402080A3EF6B3600000000001030307) MARK=0x40000
[94416.421652] nf_ct_proto_6: invalid packet ignored in state SYN_RECV  IN= OUT= SRC=192.168.219.79 DST=192.168.219.77 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=0 DF PROTO=TCP SPT=8080 DPT=3663 SEQ=678911481 ACK=4068803589 WINDOW=64308 RES=0x00 ACK SYN URGP=0 OPT (020405820402080AC88C1AF63EF6AF4B01030307) MARK=0x40000
[94418.469510] nf_ct_proto_6: invalid packet ignored in state SYN_RECV  IN= OUT= SRC=192.168.219.79 DST=192.168.219.77 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=0 DF PROTO=TCP SPT=8080 DPT=3663 SEQ=678911481 ACK=4068803589 WINDOW=64308 RES=0x00 ACK SYN URGP=0 OPT (020405820402080AC88C22F63EF6AF4B01030307) MARK=0x40000
[94418.469558] nf_ct_proto_6: invalid packet ignored in state SYN_RECV  IN= OUT= SRC=192.168.219.77 DST=192.168.219.79 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=7878 DF PROTO=TCP SPT=5566 DPT=8080 SEQ=4068803588 ACK=0 WINDOW=64860 RES=0x00 SYN URGP=0 OPT (020405820402080A3EF6BB600000000001030307) MARK=0x40000
[94422.502506] nf_ct_proto_6: invalid packet ignored in state SYN_RECV  IN= OUT= SRC=192.168.219.77 DST=192.168.219.79 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=7879 DF PROTO=TCP SPT=5566 DPT=8080 SEQ=4068803588 ACK=0 WINDOW=64860 RES=0x00 SYN URGP=0 OPT (020405820402080A3EF6CB210000000001030307) MARK=0x40000
[94422.502654] nf_ct_proto_6: invalid packet ignored in state SYN_RECV  IN= OUT= SRC=192.168.219.79 DST=192.168.219.77 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=0 DF PROTO=TCP SPT=8080 DPT=3663 SEQ=678911481 ACK=4068803589 WINDOW=64308 RES=0x00 ACK SYN URGP=0 OPT (020405820402080AC88C32B73EF6AF4B01030307) MARK=0x40000
```
But the strange thing is：if I access to RS firstly and then VS, it works. both connections are successfully established：
```
# In RS-bash:
root@ep:~# netstat -putan
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address Foreign Address State PID/Program name
tcp6 0 0 :::8080 :::* LISTEN 1/./usr/bin/main
tcp6 0 0 192.168.219.79:8080 192.168.219.77:22284 ESTABLISHED 1/./usr/bin/main
tcp6 0 0 192.168.219.79:8080 192.168.219.77:5566 ESTABLISHED 1/./usr/bin/main

#  On master:

[NEW] tcp 6 120 SYN_SENT src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 [UNREPLIED] src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=5566
[UPDATE] tcp 6 59 SYN_RECV src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=5566
[UPDATE] tcp 6 432000 ESTABLISHED src=192.168.219.77 dst=192.168.219.79 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=5566 [ASSURED]

[NEW] tcp 6 120 SYN_SENT src=192.168.219.77 dst=10.108.122.180 sport=5566 dport=8080 [UNREPLIED] src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=22284
[UPDATE] tcp 6 60 SYN_RECV src=192.168.219.77 dst=10.108.122.180 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=22284
[UPDATE] tcp 6 432000 ESTABLISHED src=192.168.219.77 dst=10.108.122.180 sport=5566 dport=8080 src=192.168.219.79 dst=192.168.219.77 sport=8080 dport=22284 [ASSURED]
```
I'm so confused. I agree this is kernel bug instead of `kube-proxy`, I don't know what's the root cause.

 *NOTE*: mode iptables is works well.
